plugins/saveactions.c

```diff
@@ -258,49 +271,322 @@ static void backupcopy_document_save_cb(GObject *obj, GeanyDocument *doc, gpoint
+	g_free(msg);
+
+	switch (response)
+	{
+		case GTK_RESPONSE_YES:
+			old_file_path = g_strdup(file_path);
+
+			if (dialogs_show_save_as())
+			{
+				new_file_path = doc->file_name;
+
+				if (! g_str_equal(old_file_path, new_file_path))
+				{
+				    //remove temp file if it was saved as some other file
+					g_remove(locale_file_path);
+				}
+			}
+			else
+			{
+				plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			}
+
+			g_free(old_file_path);
+			break;
+
+		case GTK_RESPONSE_NO:
+			g_remove(locale_file_path);
+
+			ui_set_statusbar(TRUE, _("Temp file %s was deleted"), short_filename);
+			break;
+
+		case GTK_RESPONSE_CANCEL:
+		default:
+			plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			break;
+	}
+}
+
+
+static void persistent_temp_files_document_close_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	gchar *file_path, *locale_file_path, *short_filename;
 
-		if (ft != NULL && ft->id == GEANY_FILETYPES_NONE)
-			document_set_filetype(doc, filetypes_lookup_by_name(instantsave_default_ft));
+	file_path = doc->file_name;
 
-		/* force saving the file to enable all the related actions(tab name, filetype, etc.) */
-		document_save_file(doc, TRUE);
-    }
+	if (enable_persistent_temp_files && file_path != NULL && ! geany_is_quitting())
+	{
+		if (is_temp_saved_file_doc(doc))
+		{
+			short_filename = document_get_basename_for_display(doc, -1);
+			locale_file_path = utils_get_locale_from_utf8(file_path);
+
+			if (! document_is_empty(doc))
+			{
+				show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+					doc,
+					file_path,
+					locale_file_path,
+					short_filename
+				);
+			}
+			else
+			{
+				g_remove(locale_file_path);
+
+				ui_set_statusbar(TRUE, _("Empty temp file %s was deleted"), short_filename);
+			}
+
+			g_free(short_filename);
+			g_free(locale_file_path);
+		}
+	}
+}
+
+
+static void persistent_temp_files_document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (enable_persistent_temp_files)
+	{
+		gchar *old_file_path, *new_file_path, *old_file_name;
+		GtkWidget *current_page, *page_label;
+		GtkNotebook *notebook;
+		gint current_page_num;
+
+		notebook = GTK_NOTEBOOK(geany->main_widgets->notebook);
+		current_page_num = gtk_notebook_get_current_page(notebook);
+
+		current_page = gtk_notebook_get_nth_page(notebook, current_page_num);
+		page_label = gtk_notebook_get_tab_label(notebook, current_page);
+
+		old_file_path = gtk_widget_get_tooltip_text(page_label);
+		new_file_path = DOC_FILENAME(doc);
+
+		if (old_file_path == NULL)
+		{
+			ui_set_statusbar(TRUE, _("plugin error: failed to delete initial temp file "
+				"('failed to get notebook tab label')"));
+
+			return;
+		}
+
+		old_file_name = g_path_get_basename(old_file_path);
+
+		if (is_temp_saved_file(old_file_name) && ! g_str_equal(old_file_path, new_file_path))
+		{
+			g_hash_table_insert(
```

Comment:
we have to use this hack with hashtable to remember name of original 'temp' file, during performing 'save as' which assigns a new 'final' name to file.

Ideally we would like to e.g. add 'original name' field to GeanyDocument structure but this wont happen for single plugin needs